doc/developer/design/20240610_unified_compute_introspection.md

```diff
@@ -0,0 +1,134 @@
+
+We propose extending the `CollectionManager`'s API to also accept deletion requests.
+A deletion request contains a collection ID, as well as a `filter` closure `Box<dyn Fn(&Row) -> bool>`.
+The `CollectionManager` handles a deletion request by reading back the target collection at its latest readable time, applying the `filter` closure to each read `Row`, and appending retractions for each `Row` the `filter` returned `true` for.
+Append and deletion requests must be applied by the `CollectionManager` in the same order they have been issued by the client, to guarantee that deletions will retract all previously sent appends.
+
+## Minimal Viable Prototype
+
+Prototype implementation: https://github.com/MaterializeInc/materialize/pull/27491
+
+## Alternatives
+
+### Defining Introspection Subscribes in the Coordinator
+
+It is possible to implement the unified introspection mechanism outside of the compute controller.
+The compute controller's API should be sufficient for this:
+It allows defining replica-targeted subscribes, returns subscribe responses, and interrupts replica-targeted subscribes on replica disconnections.
+Thus, an alternative design could be to let the coordinator install the introspection subscribes and act on their responses by writing them to storage collections.
+
+This alternative has the desirable properties that introspection subscribes could be defined in SQL, rather than LIR, and that they could read from arbitrary builtin relations, not just introspection indexes.
+
+The main drawback of this alternative appears to be complexity.
+During an (unfinished) attempt to develop a prototype of this approach, the amount of changes and new edge cases required in the coordinator code was significantly higher than with the design proposed above.
+The coordinator makes a number of assumptions about subscribe handling (e.g., every active subscribe is associated with a client connection) that don't hold for introspection subscribes and would need to be adjusted.
+
+In the interest of moving fast we opt to not follow this approach for the MVP implementation.
+If we find that the limitations of the controller-based approach are too great, we can revisit this decision.
+
+### Controller `CollectionId`s
+
+As an alternative to sharing the `transient_id_gen` between the coordinator and the compute controller, we could change the compute (and perhaps storage) layer to identify collections not by `GlobalId` but by a separate `CollectionId` type:
+
+```rust
+enum CollectionId {
+    External(GlobalId),
+    Internal(u64),
+}
+```
+
+This would allow the compute controller to mint its own collection identifiers without needing to synchronize with the coordinator.
+
+While this would be a desirable change, also for a possible `ALTER MATERIALIZED VIEW ... SET CLUSTER` feature, the required refactoring work would be significant.
+Again in the interest of moving fast, we opt for the simpler approach of sharing the `transient_id_gen` instead.
```

Comment:
(Already discussed and approved on Slack, but recording here for posterity my 👍🏽 for this decision.)